### PR TITLE
Add a variant of CPO, SimPO

### PR DESCRIPTION
**Pull request by fe1ixxu**
This PR adds a variant of CPO reference-free method, called SimPO (https://arxiv.org/abs/2405.14734). The only difference is SimPO adds a reward margin and does not use BC regularization. 

Thanks for your help and notice!

cc @kashif @lewtun

### CODE REVIEW

File: trl/trainer/cpo_trainer.py [hunk 1]
```diff
@@ -691,12 +703,16 @@ def cross_entropy_loss(logits, labels):
...
+        if self.loss_type != "simpo":
+            nll_loss = cross_entropy_loss(all_logits[:len_chosen], labels[:len_chosen])
+        else:
+            nll_loss = torch.tensor(0.0).to(self.accelerator.device)
 
         all_logps = self.get_batch_logps(
             all_logits,
             concatenated_batch["concatenated_labels"],
-            average_log_prob=self.loss_type == "ipo",
+            average_log_prob=self.loss_type == "ipo" or self.loss_type == "simpo",
```
Comment (kashif) [inline trl/trainer/cpo_trainer.py:715]:
```suggestion
            average_log_prob=self.loss_type in ["ipo", "simpo"],
```

**Comment by kashif** [comment]
@fe1ixxu could you kindly run `pre-commit` to fix the formatting?

**Comment by fe1ixxu** [comment]
> @fe1ixxu could you kindly run `pre-commit` to fix the formatting?

Hi @kashif,  formatting has been fixed!

**Comment by kashif** [comment]
@fe1ixxu would it be possible for you to also add a line in the cpo tests to test this loss?

**Comment by fe1ixxu** [comment]
> @fe1ixxu would it be possible for you to also add a line in the cpo tests to test this loss?

Hi @kashif, it has been added!

File: docs/source/cpo_trainer.mdx [hunk 2]
```diff
@@ -5,6 +5,9 @@ avoid generating adequate, but not perfect translations in Machine Translation (
 
 CPO aims to mitigate two fundamental shortcomings of SFT. First, SFT’s methodology of minimizing the discrepancy between predicted outputs and gold-standard references inherently caps model performance at the quality level of the training data. Secondly, SFT lacks a mechanism to prevent the model from rejecting mistakes in translations. The CPO objective is derived from the DPO objective.
 
+## A Variant of CPO: SimPO
+There is also a variant of CPO, [SimPO: Simple Preference Optimization with a Reference-Free Reward](https://arxiv.org/abs/2405.14734), which adds a reward margin and does not use BC regularization.
```
Comment (kashif) [inline docs/source/cpo_trainer.mdx:9]:
```suggestion
There is also a variant of CPO, [SimPO: Simple Preference Optimization with a Reference-Free Reward](https://arxiv.org/abs/2405.14734), which adds a reward margin and does not use BC regularization. Use the `loss_type="simpo"` in the `CPOConfig` to use this loss.
```

File: trl/trainer/cpo_config.py [hunk 3]
```diff
@@ -64,8 +64,9 @@ class CPOConfig(TrainingArguments):
 
     beta: float = 0.1
     label_smoothing: float = 0
-    loss_type: Literal["sigmoid", "hinge", "ipo", "kto_pair"] = "sigmoid"
+    loss_type: Literal["sigmoid", "hinge", "ipo", "kto_pair", "simpo"] = "sigmoid"
     disable_dropout: bool = True
+    simpo_gamma: float = 0.5
```
Comment (kashif) [inline trl/trainer/cpo_config.py:69]: add a doc string to for the `simpo_gamma` above

**Comment by kashif** [comment]
awesome thanks @fe1ixxu i'll let @lewtun do a review and we should be good! thanks!

**Comment by fe1ixxu** [comment]
@kashif  Thanks so much for the quick help!

File: trl/trainer/cpo_config.py (hunk 3 above)
Comment (fe1ixxu) [reply trl/trainer/cpo_config.py:69]: Added!

**Review by kashif — APPROVED** [review]

File: trl/trainer/cpo_config.py [hunk 4]
```diff
@@ -64,8 +64,12 @@ class CPOConfig(TrainingArguments):
...
+    """
+    simpo_gamma: A target reward margin for the SimPO loss, used only when the "simpo" option is enabled.
+    """
```
Comment (kashif) [inline trl/trainer/cpo_config.py:72]: @fe1ixxu  i meant perhaps move this up into the docstrings?

Comment (fe1ixxu) [reply trl/trainer/cpo_config.py]: Sorry about the misunderstanding! I have moved it to the above docstrings!

File: docs/source/cpo_trainer.mdx [hunk 5]
```diff
@@ -5,6 +5,9 @@ avoid generating adequate, but not perfect translations in Machine Translation (
 
 CPO aims to mitigate two fundamental shortcomings of SFT. First, SFT’s methodology of minimizing the discrepancy between predicted outputs and gold-standard references inherently caps model performance at the quality level of the training data. Secondly, SFT lacks a mechanism to prevent the model from rejecting mistakes in translations. The CPO objective is derived from the DPO objective.
 
+## A Variant of CPO: SimPO
+There is also a variant of CPO, [SimPO: Simple Preference Optimization with a Reference-Free Reward](https://arxiv.org/abs/2405.14734), which adds a reward margin and does not use BC regularization. Use the `loss_type="simpo"` in the `CPOConfig` to use this loss.
+
```
Comment (kashif) [inline docs/source/cpo_trainer.mdx:10]: replace by:

```md
## SimPO: Normalizing output length

[SimPO: Simple Preference Optimization with a Reference-Free Reward](https://arxiv.org/abs/2405.14734) is an alternative loss that adds a reward margin, allows for length normalization, and does not use BC regularization. We can use SimPO easily by turning on`loss_type="simpo"` in the `CPOConfig` to use this loss.
```

**Review by vwxyzjn — APPROVED** [review]

**Comment by fe1ixxu** [comment]
Thanks for the help! I appreciate it! @vwxyzjn @kashif
